classes/userGroup/Repository.php

```diff
@@ -0,0 +1,469 @@
+    {
+        $schemaService = Services::get('schema');
+
+        $validator = ValidatorFactory::make(
+            $props,
+            $schemaService->getValidationRules(PKPSchemaService::SCHEMA_USER_GROUP, $allowedLocales)
+        );
+
+        // Check required fields
+        ValidatorFactory::required(
+            $validator,
+            $userGroup,
+            $schemaService->getRequiredProps(PKPSchemaService::SCHEMA_USER_GROUP),
+            $schemaService->getMultilingualProps(PKPSchemaService::SCHEMA_USER_GROUP),
+            $allowedLocales,
+            $primaryLocale
+        );
+
+        // Check for input from disallowed locales
+        ValidatorFactory::allowedLocales($validator, $schemaService->getMultilingualProps(PKPSchemaService::SCHEMA_USER_GROUP), $allowedLocales);
+
+        $errors = [];
+        if ($validator->fails()) {
+            $errors = $schemaService->formatValidationErrors($validator->errors());
+        }
+
+        HookRegistry::call('UserGroup::validate', [$errors, $userGroup, $props, $allowedLocales, $primaryLocale]);
+
+        return $errors;
+    }
+
+    /**
+     * @copydoc \PKP\services\entityProperties\EntityWriteInterface::add()
+     */
+    public function add(UserGroup $userGroup): int
+    {
+        $userGroupId = $this->dao->insert($userGroup);
+        $userGroup = Repo::userGroup()->get($userGroupId);
+
+        HookRegistry::call('UserGroup::add', [$userGroup]);
+
+        return $userGroup->getId();
+    }
+
+    /**
+     * @copydoc \PKP\services\entityProperties\EntityWriteInterface::edit()
+     */
+    public function edit(UserGroup $userGroup, array $params)
+    {
+        $newUserGroup = Repo::userGroup()->newDataObject(array_merge($userGroup->_data, $params));
+
+        HookRegistry::call('UserGroup::edit', [$newUserGroup, $userGroup, $params]);
+
+        $this->dao->update($newUserGroup);
+
+        Repo::userGroup()->get($newUserGroup->getId());
+    }
+
+    /**
+     * @copydoc \PKP\services\entityProperties\EntityWriteInterface::delete()
+     */
+    public function delete(UserGroup $userGroup)
+    {
+        HookRegistry::call('UserGroup::delete::before', [$userGroup]);
+
+        $this->dao->delete($userGroup);
+
+        HookRegistry::call('UserGroup::delete', [$userGroup]);
+    }
+
+    /**
+    * Delete all user groups assigned to a certain context by contextId
+    *
+    * @param int $contextId
+    */
+    public function deleteByContextId($contextId)
+    {
+        // I may have to use transactions here
+        $collector = Repo::userGroup()->getCollector()->filterByContextIds([$contextId]);
+        $userGroupIds = Repo::userGroup()->getIds($collector);
+        foreach ($userGroupIds as $userGroupId) {
+            $this->dao->deleteById($userGroupId);
+        }
+    }
+
+    /**
+    * Get all user groups assigned to a certain context by contextId
+    */
+    public function getByContextId(int $contextId)
```

Comment:
I'm on the fence about whether to remove these functions now, in favour of consistently using `Collector`s in calling code. Since we're moving this function to a different class, this is a breaking change, so it makes sense to get rid of them now. What do you think?

Reply:
I had the same question when I wrote it. It makes sense to remove it.

I retained it because they are being used in a lot of places in the code (33 hits in 21 files), thinking that it would be better to contain the change in a certain function. (especially before being sure that it works as it should).

Reply:
Changed - `getByContextId` function removed